File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/hardware/physical/request/ContinousServoRequest.java
package org.firstinspires.ftc.teamcode.hardware.physical.request;

import com.michaell.looping.ScriptParameters;
import com.qualcomm.robotcore.hardware.CRServo;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.util.Range;
import org.firstinspires.ftc.teamcode.hardware.physical.inputs.ContinousServoInput;
import org.firstinspires.ftc.teamcode.hardware.physical.inputs.ContinousServoOptions;

public class ContinousServoRequest extends ScriptParameters.Request {

    private final CRServo SERVO;

    public ContinousServoRequest(String name, HardwareMap hardwareMap) {
        super(name);
        SERVO = hardwareMap.get(CRServo.class, name);
        SERVO.resetDeviceConfigurationForOpMode();
        SERVO.setDirection(DcMotorSimple.Direction.FORWARD);
    }

    @Override
    public Object issueRequest(Object o) {
        ContinousServoInput input = (ContinousServoInput) o;
        if(input.getType() == ContinousServoOptions.GET) {
            return SERVO.getPower() * 100;
        }else{
            double val = input.getPower();
            val = Range.clip(val, -100, 100);
            SERVO.setPower(val / 100);
            return 0;
        }
    }

    @Override
    public Class getOutputType() {
        return double.class;
    }

    @Override
    public Class getInputType() {
        return ContinousServoInput.class;
    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/internals/hardware/requests/ContinousServoRequest.kt
package org.firstinspires.ftc.teamcode.internals.hardware.requests

import com.michaell.looping.ScriptParameters
import com.qualcomm.robotcore.hardware.CRServo
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.HardwareMap
import com.qualcomm.robotcore.util.Range
import org.firstinspires.ftc.teamcode.internals.hardware.data.ContinousServoInput
import org.firstinspires.ftc.teamcode.internals.hardware.data.ContinousServoOptions

class ContinousServoRequest(name: String, hardwareMap: HardwareMap) : ScriptParameters.Request(name) {
    val servo: CRServo

    init {
        servo = hardwareMap.get(CRServo::class.java, name)
        servo.resetDeviceConfigurationForOpMode()
        servo.direction = DcMotorSimple.Direction.FORWARD
    }

    override fun issueRequest(o: Any): Any {
        val input = o as ContinousServoInput
        return if (input.type === ContinousServoOptions.GET) {
            servo.power * 100
        } else {
            var `val` = input.power
            `val` = Range.clip(`val`, -100.0, 100.0)
            servo.power = `val` / 100
            0
        }
    }

    override fun getOutputType(): Class<*>? {
        return Double::class.javaPrimitiveType
    }

    override fun getInputType(): Class<*> {
        return ContinousServoInput::class.java
    }
}